File: 0713-subarray-product-less-than-k/0713-subarray-product-less-than-k.cpp
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        int i = 0, j = 0, n = nums.size(), ans = 0;
        long long prod = 1;
        while(j < n){
            prod = prod * nums[j];
            while(prod >= k and i <= j){
                prod = prod / nums[i];
                ++i;
            }
            ans = ans + (j - i + 1);
            ++j;
        }
        return ans;
    }
};

File: 0713-subarray-product-less-than-k/0713-subarray-product-less-than-k.java
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        int prod = 1;
        int i = 0, j = 0, n = nums.length;
        int res = 0;
        
        while(j < n){
            prod = prod * nums[j];
            
            while(i < j && prod >= k){
                prod = prod / nums[i];
                ++i;
            }
            
            if(prod < k)    
                res += j - i + 1;
            
            ++j;
        }
        
        return res;
    }
}